dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1022 @@
+  want instead of at start time (or just never altogether if the return doesn't matter to the user).
+* For string-based calls to `StartWorkflowAsync`, we are accepting all args at once via `object[]`. But since array
+  types are covariant in .NET, this means you could have a workflow that accepts `string[]` and forget that needs to
+  just be the first value of the object array.
+  * Cannot think of a good way to make this better. We could:
+    * Use varargs (i.e. `params`), but not having options as the last parameter is ugly DX
+    * Use some kind of forced wrapper, e.g. `IArgumentCollection`, but it's ugly too
+    * Accept args in options instead of as param, but that options set is reused for other calls
+    * Accept both single-arg and multi-arg w/ "unset" defaults, but that is confusing and most .NET devs use their
+      positional parameters unnamed
+    * Have separate overloads for no-arg, 1-arg, and multi-arg but that's not any clearer
+* For `GetWorkflowHandle`, we cannot accept a workflow reference because in .NET if you don't type all type arguments
+  (i.e. including that parameter type), you have to help the type inferencer out. So since they'd have to provide a type
+  argument anyways, we make it where they either can here or wait until `GetResultAsync` to do so.
+  * We could have taken a proxy approach, but that would have made the code look like
+    `client.GetWorkflowHandle<GreetingWorkflow>("my wf id", wf => wf.RunAsync)` which is not much better and is hard to
+    get right with all of the overloads so you can have the lambda as the last param. This is what Azure Durable
+    Entities does.
+* Cancellation tokens (and RPC metadata and RPC timeout and such) will be in the call options, not as the last parameter
+  in a bunch of overloads like many other .NET methods
+* Older versions of .NET don't have `IAsyncEnumerator` and we don't want to go out of our way to provide a special
+  `ListWorkflows` for them
+  * Older versions can just use raw gRPC or not have this feature. These older .NET versions are really old and we don't
+    expect hardly any users on them.
+  * Note, `ListWorkflows` is not `ListWorkflowsAsync` because the first page is lazily fetched on first iteration
+  * `WorkflowExecutionEnumerator` has a `MapHistories` on it that returns `IAsyncEnumerator<WorkflowHistory>` (or is an
+    extension method)
+
+### WorkflowHandle
+
+Looks like:
+
+```csharp
+namespace Temporal.Client;
+
+public class WorkflowHandle<TResult>
+{
+    public string ID { get; }
+    public string? RunID { get; }
+    public string? ResultRunID { get; }
+    public string? FirstExecutionRunID { get; }
+
+    public async Task<TResult> GetResultAsync(
+        bool followRuns = true, RPCOptions? options = null) { /*...*/ }
+
+    public async Task<TResultLocal> GetResultAsync<TResultLocal>(
+        bool followRuns = true, RPCOptions? options = null) { /*...*/ }
```

Comment:
Possibly consider using a marker type like `Untyped`, and only adding this method to `WorkflowHandle<Untyped>` via an extension method. (And add an explicit `client.GetUntypedWorkflowHandle` which returns that)

That way there's an explicit way of creating an untyped workflow handle, and you avoid being able to specify a different result type when getting the result than was specified when you created the handle, which ideally is a state that can't be represented.

Reply:
I can make this an extension method, Will think on it. I can also provide something like https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.cast. I like the simplicity of having it here for discoverability and for anyone to force-ably override the result type, but will ponder.